File: api/src/main/java/xtdb/IllegalArgumentException.java
package xtdb;

import clojure.lang.IExceptionInfo;
import clojure.lang.IPersistentMap;
import clojure.lang.Keyword;

import java.util.Objects;

@SuppressWarnings("unused")
public class IllegalArgumentException extends java.lang.IllegalArgumentException implements IExceptionInfo {
    private static final long serialVersionUID = 8569715935234823692L;
    private static final Keyword ILLEGAL_ARGUMENT = Keyword.intern("illegal-argument");
    private static final Keyword ERROR_TYPE = Keyword.intern("xtdb.error", "error-type");
    private static final Keyword ERROR_KEY = Keyword.intern("xtdb.error","error-key");
    private static final Keyword MESSAGE_KEY = Keyword.intern("xtdb.error", "message");

    private final IPersistentMap data;

    public IllegalArgumentException(String message, IPersistentMap data, Throwable cause) {
        super(message, cause);
        this.data = data;
    }

    @Override
    public IPersistentMap getData() {
        return data;
    }

    public static IllegalArgumentException create(Keyword error, String message, IPersistentMap data) {
        return create(error, message, data,null);
    }
    public static IllegalArgumentException create(Keyword error, String message, IPersistentMap data, Throwable cause) {
        message = message != null ? message : (String) data.valAt(MESSAGE_KEY, String.format("Illegal argument: '%s'", error.toString()));
        return new IllegalArgumentException(message, data.assoc(ERROR_TYPE, ILLEGAL_ARGUMENT).assoc(ERROR_KEY, error).assoc(MESSAGE_KEY, message), cause);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        var that = (IllegalArgumentException) o;
        return Objects.equals(getMessage(), that.getMessage()) && Objects.equals(getCause(), that.getCause()) && Objects.equals(data, that.data);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getMessage(), getCause(), data);
    }

    @Override
    public String toString() {
        return String.format("IllegalArgumentException{message=%s, data=%s}", getMessage(), data);
    }
}

File: api/src/main/kotlin/xtdb/IllegalArgumentException.kt
package xtdb

import clojure.lang.IExceptionInfo
import clojure.lang.IPersistentMap
import clojure.lang.Keyword
import clojure.lang.PersistentHashMap
import clojure.lang.Util

private val ERROR_KEY: Keyword = Keyword.intern("xtdb.error", "error-key")

/**
 * @suppress
 */
@Suppress("unused")
data class IllegalArgumentException(
    val key: Keyword?,
    override val message: String? = "Illegal argument: '${key?.sym}'",
    private val data: Map<*, *> = emptyMap<Keyword, Any>(),
    override val cause: Throwable? = null,
) : java.lang.IllegalArgumentException(message, cause), IExceptionInfo {

    override fun getData(): IPersistentMap =
        (data as? IPersistentMap ?: PersistentHashMap.create(data)).assoc(ERROR_KEY, key)

    // exclude `cause`
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is IllegalArgumentException) return false

        if (key != other.key) return false
        if (message != other.message) return false
        if (!Util.equiv(data, other.data)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = key?.hashCode() ?: 0
        result = 31 * result + (message?.hashCode() ?: 0)
        result = 31 * result + Util.hash(data)
        return result
    }

    /**
     * @suppress
     */
    companion object {
        @JvmStatic
        @JvmOverloads
        fun createNoKey(message: String, data: Map<*, *>, cause: Throwable? = null) =
            IllegalArgumentException(null, message, data, cause)

        @JvmStatic
        @JvmOverloads
        fun create(key: Keyword?, data: Map<*, *>, cause: Throwable? = null) =
            IllegalArgumentException(key, data = data, cause = cause)
    }


}
